src/com/google/javascript/jscomp/CheckGlobalThis.java

```diff
@@ -99,6 +101,14 @@ public boolean shouldTraverse(NodeTraversal t, Node n, Node parent) {
         return false;
       }
 
+      if (n.getJSType() instanceof FunctionType) {
+        FunctionType functionType = (FunctionType)n.getJSType();
+        if (functionType.getTypeOfThis() != compiler.getTypeRegistry().getNativeType(JSTypeNative.UNKNOWN_TYPE))
```

Comment:
I'm not sure this does what we want.

Could you add test cases to CheckGlobalThisTest to be sure?
Thanks.

Reply:
Thanks for the feedback. Here is what I have done to test:

Content of pass.js
```js
/**
 * @type {function(this:{hello:string})}
 */
function test() {
  console.log(this.hello)
}

test.call({ hello: 'world' })
```

Content of fail.js
```js
/**
 * @type {function()}
 */
function test() {
  console.log(this.hello)
}

test.call({ hello: 'world' })
```

The goal of the issue, as I understand, is to get `pass.js` to pass and `fail.js` to fail. Here I am able to achieve the goal:

I ran:
`java -jar target/closure-compiler-1.0-SNAPSHOT.jar --compilation_level ADVANCED pass.js`

And the result is:
```js
console.log({a:"world"}.a);
```

Now I ran:
`java -jar target/closure-compiler-1.0-SNAPSHOT.jar --compilation_level ADVANCED fail.js`

And now the result is:
```
fail.js:5: WARNING - [JSC_USED_GLOBAL_THIS] dangerous use of the global this object
  console.log(this.hello)
              ^^^^

0 error(s), 1 warning(s), 83.3% typed
console.log({a:"world"}.a);
```

I was wondering about your comments in the issue around the `To fix this bug for the case where type checking is disabled` clause, so I attempted to disable type checking.

To my surprise, specifying `--use_type_for_optimization=false` doesn't change my experiment result, the pass.js still pass and fail.js still fail.

Without more information available, I just opened the pull request, thinking it is sufficient. I attempted to add this unit test and now it fails.

```Java
  public void testPass()
  {
    testSame(lines(
      "/**",
      " * @type {function(this:{hello:string})}",
      " */",
      "function test() {",
      "  console.log(this.hello)",
      "}"
    ));
  }
``` 

The test case failed because the `n.getJSType()` is null and so the logic to skip raising the warning is not triggered. This is probably the case you were talking about that I was unable to trigger.

Moving forward, it would be interesting to see how I can leverage the `jsdoc` available in line 95 to perform essentially the same check.

Reply:
Could you add your test cases to this file?

https://github.com/google/closure-compiler/blob/master/test/com/google/javascript/jscomp/CheckGlobalThisTest.java

Thanks.